kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/SchemaUtils.java

```diff
@@ -79,6 +82,114 @@ public static void validateSchema(StructType schema, boolean isColumnMappingEnab
     validateSupportedType(schema);
   }
 
+  /**
+   * Performs the following validations on an updated table schema:
+   *
+   * <p>The following checks are performed:
+   *
+   * <ul>
+   *   <li>No duplicate columns are allowed
+   *   <li>Names contain only valid characters
+   *   <li>Data types are supported
+   *   <li>No new non-nullable fields are added
+   *   <li>Physical column name consistency is preserved in the new schema
+   *   <li>Partition columns are not modified
+   *   <li>Nested IDs for array/map types are preserved in the new schema
+   *   <li>No type changes
+   * </ul>
+   */
+  public static void validateUpdatedSchema(
+      StructType currentSchema,
+      StructType newSchema,
+      List<String> partitionCols,
+      Metadata metadata) {
+    validateSchema(newSchema, true /*columnMappingEnabled*/);
+    Map<String, StructField> currentColumnPathToField = columnPathToField(currentSchema);
+    Map<String, StructField> updatedColumnPathToField = columnPathToField(newSchema);
+    ColumnMapping.validateColumnIds(
+        currentColumnPathToField,
+        updatedColumnPathToField,
+        TableConfig.ICEBERG_COMPAT_V2_ENABLED.fromMetadata(metadata));
+    validateUpdatedSchemaCompatibility(
+        newSchema, currentColumnPathToField, updatedColumnPathToField, partitionCols);
+  }
+
+  /* Verifies that no non-nullable fields are added, no type changes are performed and that
+  partition columns are not modified */
+  private static void validateUpdatedSchemaCompatibility(
+      StructType newSchema,
+      Map<String, StructField> currentColumnPathToStruct,
+      Map<String, StructField> updatedColumnPathToStruct,
+      List<String> partitionCols) {
+    validatePartitionColumns(newSchema, partitionCols);
+    for (Map.Entry<String, StructField> newFieldEntry : updatedColumnPathToStruct.entrySet()) {
+      if (!currentColumnPathToStruct.containsKey(newFieldEntry.getKey())) {
+        if (!newFieldEntry.getValue().isNullable()) {
+          throw new KernelException(
+              String.format(
+                  "Cannot add a non-nullable field %s", newFieldEntry.getValue().getName()));
+        }
+      } else {
+        StructField newField = newFieldEntry.getValue();
+        StructField existingField = currentColumnPathToStruct.get(newFieldEntry.getKey());
+        validateNoTypeChange(existingField, newField);
+      }
```

Comment:
Checking type change and nullabiility can probably be done in the same pass over the field, I think we can clean this up later though